res/controllers/common-hid-utils.js

```diff
@@ -0,0 +1,83 @@
+HIDController.prototype.defaultScalers = {
+    'volume': function (value) { return script.absoluteLin(value, 0, 1, 0, 4096) },
+    'volumefader': function (value) { return Math.pow(HIDController.prototype.defaultScalers.volume(value), 2) },
+    'eq': function (value) { return script.absoluteLin(value, 0, 2, 0, 4096) },
+    'pregain': function (value) { return script.absoluteNonLin(value, 0, 1, 4, 0, 4096) },
+    'rate': function (value) { return script.absoluteLin(value, -1, 1, 0, 4096) },
+    'plusminus': function (value) { return script.absoluteLin(value, -1, 1, -4096, 4096) }
+}
+
+HIDController.prototype.softTakeover = function() {
+    for(var i = 1; i < arguments.length; i++) {
+        engine.softTakeover(arguments[0], arguments[i], true);
+    }
+}
+
+HIDController.prototype.softTakeoverAll = function() {
+    for (var channel = 1; channel < 5; channel++) {
+        this.softTakeover('[EqualizerRack1_[Channel' + channel + ']_Effect1]', 'parameter1', 'parameter2', 'parameter3')
+        this.softTakeover('[QuickEffectRack1_[Channel' + channel + ']]', 'super1')
+        this.softTakeover('[Channel' + channel + ']', 'pregain', 'volume', 'rate')
+        for (var i = 1; i < 4; i++) {
+            this.softTakeover('[EffectRack1_EffectUnit' + channel + '_Effect' + i + ']', 'meta')
+        }
+    }
+    this.softTakeover('[Master]', 'headMix', 'crossfader')
+}
+
+HIDController.prototype.clearLights = function(packetName) {
+    const groups = this.getLightsPacket(packetName).groups
+    for (var groupName in groups) {
+        const group = groups[groupName]
+        for (var control in group) {
+            group[control].value = 0
+        }
+    }
+    this.sendLightsUpdate()
+}
+
+HIDController.prototype.sendLightsUpdate = function (packetName) {
+    this.getLightsPacket(packetName).send()
+}
+
+HIDController.prototype.getLightsPacket = function (packetName) {
+    return this.getOutputPacket(packetName ? packetName : 'lights')
+}
+
+
+HIDController.prototype.connectLight = function (group, name, setter) {
```

Comment:
It sets the initial value and then on every update automatically sends the new lights packet.

Reply:
`HIDController.prototype.linkOutput` already does that too. As far as I can tell the only difference is that you need to manually send the packet once after all calls to `linkOutput` because it sets the state of the packet without sending it.

Reply:
I don't quite understand how `linkOutput` operates. We definitely need more documentation on the common methods as well as the engine methods.

Reply:
`HIDController.linkOutput` is odd because it takes two pairs of "groups" and "names". The first pair refers to a name for a field which is only used by the common-hid-packet-parser.js library and can be different from the ControlObject name in Mixxx. The second pair is used to connect the callback function to changes in the state of a ControlObject. The two pairs [can be the same](https://github.com/mixxxdj/mixxx/blob/33fa74510409e94404b682047437e353a1c7d13b/res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js#L560)... which I think that is overcomplicated and confusing, but so is the whole library. IMO adding this new `connectLight` function which does largely the same thing as the existing function creates an even more confusing situation because it's not clear which function should be used.